### ISSUE TITLE
Improve error if `BoxDecoration` has both a circle shape and a border radius

### ISSUE THREAD

**Issue by loic-sharma** [flutter/flutter#158207]
### Use case

The following `BoxDecoration` is invalid:

```dart
BoxDecoration(
  shape: BoxShape.circle,
  borderRadius: BorderRadius.circular(3.0),
)
```

Using this decoration in a `DecoratedBox` widget will result in errors like:

```
════════ Exception caught by rendering library ═════════════════════════════════
'package:flutter/src/painting/box_decoration.dart': Failed assertion: line 419 pos 16: '_decoration.borderRadius == null': is not true.
The relevant error-causing widget was:
    DecoratedBox ContainerDecorators|decoratedBox:file:///Users/loicsharma/Code/tmp/decorators_study/lib/sunflower/decorator.dart:121:8
The following RenderObject was being processed when the exception was fired: RenderDecoratedBox#a007e relayoutBoundary=up2
RenderObject: RenderDecoratedBox#a007e relayoutBoundary=up2
════════════════════════════════════════════════════════════════════════════════
````

It's not clear from this error what needs to be changed.

### Proposal

Add an assertion to `BoxDecoration`'s constructor that gives a helpful error like:

```
A circle cannot have a border radius. Remove either the shape or the borderRadius argument.
```

**Comment by huycozy** [comment]
Hi @loic-sharma 
When I use the following sample code (only this widget on screen), I see the error below that has fallen to another assertion than yours:

```dart
Container(
  decoration: BoxDecoration(
    shape: BoxShape.circle,
    borderRadius: BorderRadius.circular(3.0),
  ),
  child: const Text('Click me'),
)
```

```console
'package:flutter/src/painting/box_decoration.dart': Failed assertion: line 133 pos 12: 'shape != BoxShape.circle || borderRadius == null': is not true.
```

https://github.com/flutter/flutter/blob/66e8f53ba02a16924104730e62246b0b4e072ff4/packages/flutter/lib/src/painting/box_decoration.dart#L133

I see the assertion you mentioned falls to `_paintBox` which may happen while painting something but not on a simpler case as the same code I tried?

https://github.com/flutter/flutter/blob/66e8f53ba02a16924104730e62246b0b4e072ff4/packages/flutter/lib/src/painting/box_decoration.dart#L419

So I wonder if we should also improve error message for other assertions or only the case as reported?

**Comment by loic-sharma** [comment]
Yes `Container`'s constructor calls `BoxDecoration.debugAssertIsValid`:

https://github.com/flutter/flutter/blob/c12a3dc5b3b9a19309bfcfd15efb9fb0c7b8c969/packages/flutter/lib/src/widgets/container.dart#L270

However `DecoratedBox`'s constructor does not:

https://github.com/flutter/flutter/blob/c12a3dc5b3b9a19309bfcfd15efb9fb0c7b8c969/packages/flutter/lib/src/widgets/container.dart#L67-L72

This is intentional because `DecoratedBox`'s constructor is `const` and cannot invoke methods. We could fix the `DecoratedBox` scenario by adding an assertion to `BoxDecoration`'s constructor. This would address both the `Container` and `DecoratedBox` scenarios.

**Comment by dbspoudel** [comment]
@loic-sharma I'd love to contribute a fix for this.

**Comment by loic-sharma** [comment]
@dbspoudel Go for it, it's yours! :)

**Comment by dbspoudel** [comment]
> [@dbspoudel](https://github.com/dbspoudel) Go for it, it's yours! :)

Thank you @loic-sharma 
I’ve submitted a PR https://github.com/flutter/flutter/pull/163917 and would appreciate your review. Let me know if any changes are needed.

**Comment by dbspoudel** [comment]
@loic-sharma waiting for your review. #163917
